src/adapter/src/catalog/transact.rs

```diff
@@ -458,19 +480,72 @@ impl Catalog {
         let mut storage_collections_to_drop = BTreeSet::new();
 
         for op in ops {
-            Self::transact_op(
+            let (weird_builtin_table_update, temporary_item_actions) = Self::transact_op(
                 oracle_write_ts,
                 session,
                 op,
                 &temporary_ids,
-                builtin_table_updates,
                 audit_events,
                 tx,
                 state,
                 &mut storage_collections_to_create,
                 &mut storage_collections_to_drop,
             )
             .await?;
+
+            // Certain builtin tables are not derived from the durable catalog state, so they need
+            // to be updated ad-hoc based on the current transaction. This is weird and will not
+            // work if we ever want multi-writer catalogs or high availability catalogs. If this
+            // instance crashes after committing the durable catalog but before applying the weird
+            // builtin table updates, then they'll be lost forever.
+            if let Some(builtin_table_update) = weird_builtin_table_update {
+                builtin_table_updates.push(builtin_table_update);
+            }
+
+            // Temporary items are not stored in the durable catalog, so they need to be handled
+            // separately for updating state and builtin tables. Temporary items cannot interact
```

Comment:
What does "interact" mean? A temp view can be created over non-temp items, which seems like interacting. Maybe a different word or more specifics.

Reply:
Ah, I forgot that you could create temp views over non-temp items. I was only considering creating non-temp views over temp items. In that case this comment is wrong and we probably do in fact need to weave these into the right places. I'll have to think about this. I'll also try and come up with some test cases that break down here.

Reply:
Well apparently this is already broken on `main`. I'll try and fix it on main first, so I can make sure that whatever I come up with doesn't further break it.
![Screenshot from 2024-06-18 10-37-29](https://github.com/MaterializeInc/materialize/assets/10800303/5e7b1fe4-b8b7-4b14-b8f6-e4e02e34bc04)

Reply:
https://github.com/MaterializeInc/materialize/pull/27736

Reply:
Ok, the test is passing again and temporary items should be working now. Essentially the change is:

- I added an enum variant `TemporaryItem` to `StateUpdateKind`.
- I added the temporary item updates to the large list of `updates`.
- When sorting `updates`, the temporary items are put in the correct order w.r.t. normal items.